### ISSUE TITLE
🙏 Help wanted: Deprecate old plugin backends 

### ISSUE THREAD

**Issue by jhaals** [backstage/backstage#26353]
**TL;DR** make sure that `createRouter` and other exports are marked as deprecated. For the majority of packages there should only be one default export of the backend plugin itself. The `@backstage/backend-common` package is deprecated so usages of that package should also be avoided.

Backstage’s [new backend system](https://backstage.io/docs/backend-system/) is ready for general use; we are now asking for a full transition over to the new backend system, which involves [stop supporting the old system](https://backstage.io/docs/backend-system/building-plugins-and-modules/migrating/#remove-support-for-the-old-backend-system). By old system we mean having exports of `createRouter` and related types. There should only need to be one export like [this](https://github.com/backstage/backstage/blob/5d92499845e60e154b80481a679bfa3ec12f0dfb/plugins/notifications-backend/src/index.ts#L16) in the backend plugin’s `index.ts` file.

## How do I help?

Run `yarn backstage-repo-tools lint legacy-backend-exports` in the repository to get a full report of packages that require action.

#### Ensure that plugins contain a default export

Ensure that there is a default export of the backend plugin in `index.ts`, see [this example](https://github.com/backstage/backstage/blob/5d92499845e60e154b80481a679bfa3ec12f0dfb/plugins/notifications-backend/src/index.ts#L16).

If the plugin previously had a default export in `plugins/<plugin-id>/src/alpha.ts`, make sure that that export is deprecated and that the default export is moved to the non-alpha `index.ts` instead.

#### Deprecate `createRouter`, `RouterOptions` and similar types.

Here’s an [example](https://github.com/backstage/backstage/pull/26078) of a plugin’s `createRouter` being deprecated.

```diff
/**
+ * @deprecated Please migrate to the new backend system as this will be removed in the future.
 * @public
 * */
export async function createRouter(
  options: RouterOptions,
): Promise<express.Router> {
```

#### Remove deprecated exports

Ensure that deprecations have been out for at one mainline release before proceeding with removal of all deprecated exports. Removing exports from one release to another is not recommended 

The complete migration story for a backend plugin (including deprecation) is also mentioned in our [docs](https://backstage.io/docs/backend-system/building-plugins-and-modules/migrating#remove-support-for-the-old-backend-system).

If you take on migrating a backend plugin, feel free to add a comment in this issue to avoid duplicate work.

We plan to have all `createRouter` exports and `@backstage/backend-common` usages removed by the end of this year. Your help would be much appreciated!

**Comment by avila-m-6** [comment]
Hey, I have started to work on `plugins/user-settings-backend` in #26584

**Comment by kenilkevadiya** [comment]
Hey @Rugvip, Can I work on this issue?

**Comment by itsalam** [comment]
Just working on the app-backend plugin. Since `createRootLogger` is used in much of the [test cases in the plugins](https://github.com/backstage/backstage/blob/17ecea82239f25a4ee89f9c59fe43e735d1c80a2/plugins/app-backend/src/service/appPlugin.test.ts#L34), should this be swapped out for what's found here? https://backstage.io/docs/backend-system/core-services/logger/.

**Comment by jescalada** [comment]
Hello @Rugvip, I'm working on the `catalog-backend-module-puppetdb`, which doesn't have a `/plugin` directory with the plugin we would normally export by default. It has a default export in the `alpha.ts` file as follows:

```typescript
export { default } from './module';
```
In this case, do we simply deprecate the `alpha.ts` export and add the same line to the `index.ts`? The `index.ts` file is exporting everything in the `/providers` and `/puppet` directories, and I'm not sure if we want to limit or change what gets exported to the users.

Thank you!

**Comment by knowacki23** [comment]
I believe that getting rid of `@backstage/backend-common` package usages would also reduce occurrences of Vulnerability (CVE-2024-21534)

There is a vulnerability with severity 9.3 in jsonpath-plus <= 10.0.7. See https://security.snyk.io/vuln/SNYK-JS-JSONPATHPLUS-7945884

`@backstage/backend-common` package is using `@kubernetes/client-node` package which is using vulnerable `jsonpath-plus@npm:7.2.0`

**Comment by patroswastik** [comment]
I would like to work on this

**Comment by patroswastik** [comment]
I see that the command `yarn backstage-repo-tools lint legacy-backend-exports` does is showing `❌ createRouter is NOT deprecated` for those plugins, in which createRouter has already been deprecated. I've created an issue (#27668) with additional details about this problem.

**Comment by JGoggers** [comment]
Hey, I've started to work on `@backstage/plugin-permission-node`

**Comment by j1bulbul** [comment]
Hello! I’d like to work on this. 

Was going to start migrating `plugins/techdocs-backend`, but I saw that [PR #29022](https://github.com/backstage/backstage/pull/29022) touched on it slightly while migrating search-backend — but createRouter is still present, and @backstage/backend-common is still in use.

Just checking if anyone is already working on this or if I can go ahead.

**Comment by awanlin** [comment]
Hi @j1bulbul, feel free to run with `techdocs-backend`. I have looked at this in the past and it's a bit tricky so keep that in mind when you start to work on it. 👍 

The PR you listed was focused on just Search and only touched TechDocs as there was some deprecated Search code in there that needed to be removed as part of that work.

**Comment by j1bulbul** [comment]
> Hi [@j1bulbul](https://github.com/j1bulbul), feel free to run with `techdocs-backend`. I have looked at this in the past and it's a bit tricky so keep that in mind when you start to work on it. 👍
> 
> The PR you listed was focused on just Search and only touched TechDocs as there was some deprecated Search code in there that needed to be removed as part of that work.

Okay noted, will give it a go 👍

**Comment by awanlin** [comment]
Hi @j1bulbul, I'm terribly sorry, I missed this comment on my PR - https://github.com/backstage/backstage/pull/29022#discussion_r1989474772 - when I followed up here. I'm actually going to move forward with this but if you want to try and collaborate in my PR on this happy to have another set of eyes helping me spot anything I missed or that might be wrong. Again, sorry for the confusion on my part. 😢

**Comment by j1bulbul** [comment]
Hey @awanlin, no worries at all! Happy to take a look once you’ve pushed the changes. I looked into it earlier and yeah I see what you mean about this being tricky. It’d be helpful to see how you approach it - might take on another one soon if there are still plugins that need deprecating/migrating and haven’t been picked up yet.

**Comment by Teo-ShaoWei** [comment]
Hope to gain some eyeballs please: In  `plugins/kubernetes-backend`, there seems to be a few side-effect imports of `@backstage/backend-common`, e.g. this one: 
https://github.com/backstage/backstage/blob/ab209e4964ee90de5899e8269f6255243d51e811/plugins/kubernetes-backend/src/cluster-locator/CatalogClusterLocator.test.ts#L17

I guess that some time ago, `@backstage/backend-common` had side effects, but I'm not sure if we can proceed to remove it now. As I don't have familiarity on how the side effect was used, hope to get some context from us 🙏

Together with my #30352, that would remove usage in the biggest backend plugins that users will import 🎉

I'm invested and can help to create the MR for `kubernetes-backend` too. Like #28969, we are using Trivy scan and there's this nagging vulnerability that will go away once we no longer depends on `@backstage/backend-common`.

**Comment by Ayushmore1214** [comment]
Can I try working on this issue ?

**Comment by Teo-ShaoWei** [comment]
> Can I try working on this issue ?

@Ayushmore1214 oooh awesome! 🎉 I would be interested to see this moves 🙏

As @Rugvip [says](https://github.com/backstage/backstage/issues/26353#issuecomment-2373497777):
> This is completely open to contributions 🙏

**Comment by Ayushmore1214** [comment]
Sorry for inconvenience @Teo-ShaoWei , I will be busy for next 5-7 days , If someone else likes to work on this issue till then plz assign them Thank you !

**Comment by azen1k5** [comment]
I had a look at the code yesterday. Is there somewhere more information out there how to do the migration for:

* createLegacyAuthAdapters - Migrate to use the new backend system and auth services instead.
* loggerToWinstonLogger - Migrate to use the new LoggerService instead.

the doc [1] doesn't go into details on how to do that.

[1] https://backstage.io/docs/backend-system/building-backends/migrating/#removing-backstagebackend-common

**Comment by azen1k5** [comment]
Is [1] the one for auth?

[1] https://backstage.io/docs/tutorials/auth-service-migration/#making-the-new-auth-services-available-in-createrouter

**Comment by Teo-ShaoWei** [comment]
For `loggerToWinstonLogger`, I see two options:
1. Create a local copy of `loggerToWinstonLogger.ts` within `kubernetes-backend`. (refer to #29017)
2. Craft a `logProvider` using `logger`. (refer to #28987)

I'll recommend going with option (2.). It is tailored for `createProxyMiddleware` of `'http-proxy-middleware'`, which ask for the following types:
```ts
interface LogProvider {
    log: Logger;
    debug?: Logger;
    info?: Logger;
    warn?: Logger;
    error?: Logger;
}
declare type Logger = (...args: any[]) => void;
```
But if we are wrong about option (2.), we can always go with option (1.) first and kick the can for now 😂

Since this is considered a decision point by itself that we want to deliberate further, let me create an isolated PR for this.

**Comment by Teo-ShaoWei** [comment]
> Is [1] the one for auth?
>
> [1] https://backstage.io/docs/tutorials/auth-service-migration/#making-the-new-auth-services-available-in-createrouter

@azen1k5 wooohooo good find! I was confused for a while looking at it because it seems to do nothing, except make the optional `HttpAuthService` compulsory. Actually I didn't understand it wrongly, this is indeed the state after the old backend has been torn down, e.g. inside the doc you provided:
> In order to make the new auth services available to the plugin implementation in a backwards compatible way, we use the `createLegacyAuthAdapters` helper from `@backstage/backend-common`. This helper accepts both the old and new auth services, and returns implementations for the new ones. If provided with implementations for the new services it will pass them through directly, which is what we want for the new backend system.

In that case, the right thing to do is probably to chase the type upwards and make sure we always have:
```ts
httpAuth: HttpAuthService
```
instead of leaving it optional like:
```ts
httpAuth?: HttpAuthService
```
Currently that's not true and we shouldn't just blindly type assert with:
```ts
httpAuth as HttpAuthService
```
which should cause issues down the road... ...

**Comment by azen1k5** [comment]
@Teo-ShaoWei thanks for the nice explanation. I'll give it a try

**Comment by awanlin** [comment]
Hi! Just wanted to follow up here that the only remaining Plugin that still supports the legacy backend is the Kubernetes plugin. Phase 1 was completed for this plugin but we've hit roadblocks with Phase 2:

1. My initial attempt where I hit issues with the logger: https://github.com/backstage/backstage/pull/29013
2. Freben's attempt where he resolved the logger issue but hit other items: https://github.com/backstage/backstage/pull/29015

I think before we go much further we should log a dedicated issue for this so we can get to an agreed upon path forward.

We need to decided on what was mentioned here as this is the current blocker: https://github.com/backstage/backstage/pull/29015#issuecomment-2753556148

**Comment by Teo-ShaoWei** [comment]
Thanks for the perspective @awanlin 🙏

The removal of `loggerToWinstonLogger` was indeed tricky. I for sure tried what you did as a firsthand approach haha 😂 I think @drodil realized too when he worked on #24224, thus the added comment to:
```ts
// TODO: Add 'log' to LoggerService
```

Could you take a look at #30647 and see if that resolves the issue?

EDIT: Oooh you looked at it already by the time I finished the message! 💪

**Comment by Teo-ShaoWei** [comment]
@azen1k5 have you started working on it? If yes, this is yours since you reserved it first 👍 I can contribute to review of that 💪

If you haven't started yet, then if you don't mind, I can PR what I have... This issue is an itch for me and since it's weekend I thought to try and scratch it just to feel how deep a problem it is... as I scratched it I managed to get to the end of removing `@backstage/backend-common`... hahahah 🤦

**Comment by Teo-ShaoWei** [comment]
However I wasn't able to make much headway for removing legacy backend support for the Kubernetes plugin... I'm with @awanlin that we should log a dedicated issue for that and agree on a plan 👍

**Comment by awanlin** [comment]
HI @Teo-ShaoWei, did you read my message? 

> I think before we go much further we should log a dedicated issue for this so we can get to an agreed upon path forward.
> 
> We need to decided on what was mentioned here as this is the current blocker: https://github.com/backstage/backstage/pull/29015#issuecomment-2753556148

**Comment by awanlin** [comment]
Seems that you did but not in detail. This isn't simply removing the legacy backend, please read the linked comment: https://github.com/backstage/backstage/pull/29015#issuecomment-2753556148

**Comment by Teo-ShaoWei** [comment]
Ah I did, apologies 🙇 I think you are concerned with the state of `kubernetes-backend` as a whole, is that right? I tried to look at the messages and the state of the plugin, it's indeed non-trivial.

However the removal of `@backstage/backend-common` can be done separately it seems. As in I managed to remove it while having the tests in the package pass, and `yarn tsc` being ok...

I might not understand enough of the details for sure, so it's my bad if I went off-tangent 😂

**Comment by awanlin** [comment]
If we simply wanted to remove `@backstage/backend-common` then we would have just merged https://github.com/backstage/backstage/pull/29015 but as mentioned in the comment I've linked to a few times now https://github.com/backstage/backstage/pull/29015#issuecomment-2753556148 there's more to it than that.

**Comment by Teo-ShaoWei** [comment]
Ah! Yes my bad, I didn't give the code a good read. Most of what I have is indeed in #29015 already. Got too excited to chew at the problem haha.

Not sure whether I'm misunderstood though as I'm trying to approach the issue iteratively. To me, it feels like we are crossing a big gap in one jump. Would it have been easier if we take a step closer to the cliff first before the leap to the other side?

Let me submit my PR and we look at it? If it's accepted, we can reduce the problem and at least can address #28969. Some of us are willing to slowly work on this, but our organization's Trivy can't. I'll understand if we decided to reject the PR in the end.

**Comment by awanlin** [comment]
Please, slow down and stop, you're rushing off an not clearly understanding things. If we wanted to moved forward with that PR - created by the Core Maintainers - then again we would have merged the PR I referenced. I'll copy the comment I've referenced a few times as it seems you aren't reading it:

> So I don't think that this is going to work for some use cases.
> 
> I think that it's wise for us to convert some of the extension points into serviceRefs instead, which then give us the ability to get access to the default implementations of some of the dependencies without having to export a load of things from the plugin as right now it's impossible to create a ClusterLocator outside of the plugin without copying a load of code.
> 
> internally we have a requirement that the ServiceLocator needs access to the the ClusterSupplier which I think is pretty standard, but there's no way to create your own ServiceLocator without having access to the implementation of the ClusterSupplier but with extension points this is only possible if you create your own ClusterSupplier in the same scope.
> 
> Moving these to ServiceRefs would mean that you could depend on the default implementation of the ClusterSupplier when overriding the ServiceLocator with your own.

You're changes don't address this and lead to problems for adopters. The above needs to be addressed first.

I appreciate that you want to work on this but it needs to be done correctly and in a thought out way. Any security issues with `@backstage/backend-common` can be resolved with `resolutions` this is what the vast majority of the Community has done as they wait for this work.

**Comment by Teo-ShaoWei** [comment]
Frankly I could be wrong as I'm really new to this piece of code, so just some quick intuition from me.

To me #29015 is a big change, and can be abstractly separated into three parts (A + B + C).

A = changes that doesn't affect logic at all, e.g. refactorings like removal of `@backstage/backend-common`.
B = changes in API that are small and users can fix easily, e.g. removal of the `DiscoveryService`.
C = changes in logic that result in change of existing workflow, e.g. removal of `KubernetesBuilder` and all the helper functions within, so users would need to use what the new backend offers.

I would believe that https://github.com/backstage/backstage/pull/29015#issuecomment-2753556148 is about the concern that we are not ready to make the change C, because users can't rework their existing workflow.

We are talking across each other this whole time because I guess this is what's on our mind:
@awanlin: "We didn't accept change A + B + C last time. Did you realized that the problem is this big and you haven't address the issue yet?"
@Teo-ShaoWei: "I've derived A + B. Ah yes I think we can't solve C yet... Since the existing blocker is due to C, let's commit A + B first since users will be ok. We can then isolate and solve C."

Did I understand the issue incorrectly?

**Comment by Teo-ShaoWei** [comment]
Sorry if I come off as confrontational. I really appreciate the good work of the core maintainers & the Spotify folks. Backstage is some awesome stuff!

And it's because of that, my organization want to give it a serious try, so I would also hope to do what I technically can to maximize our mutual benefits. I'm currently still inclined to believe that #30653 is such an improvement for us, but I'll be glad if I can be proven wrong. With that, I hope that the PR can get a well-thought review 🙏

**Comment by awanlin** [comment]
No worries @Teo-ShaoWei, the issue is trying to move before having an agreed upon course of action. Sometimes going slow can be fast in the end. I agree with the incremental approach, this is why I approved the Logger PR and worked to get the Kubernetes Plugin owner to approve as well. But what remains is a bit overlapping and I would prefer we get confirmation before taking further steps.

Some additional background, it's the summer time so many people from the project are away so getting this fully reviewed and merged will take time and likely won't make the next month release which is August 19th. So getting this right from the start will help make sure it moves forward with less blocking points.

**Comment by azen1k5** [comment]
@Teo-ShaoWei all good take it. I just started and reading up on the code since it's the first time for me seeing backstage from the inside. Happy to have a look at your PR to understand the ins and outs :-)

**Comment by freben** [comment]
I took a stab at the very last bits that gave errors on the command listed in the issue description. ^
